### DISCUSSION TITLE
Clarification regarding multiple workers and thread safety

### DISCUSSION THREAD

**Discussion by cayolblake** [walkor/Workerman#801] [General]
Hi @walkor 

If I have the following pseduo code.

```
use Workerman\Worker;
$w = new Worker('http://127.0.0.1:2346');
$w->count = 2;

$my_non_threadsafe_component = new abc();

$w->onMessage = function ($connection, $request) use ($my_non_threadsafe_component) {
    $my_non_threadsafe_component->write("helloworld");
    $connection->send($my_non_threadsafe_component->read("somekey"));
});

```

To my understanding, which might be wrong, $w->count should be 1, however, I did try setting the count to 2 or more, and couldn't observe any failures/issues however the performance definitely is much better. 

My question is, given that I instantiate the non-thread-safe DB instance one time, is it OKAY/SAFE to increase Workerman's thread/child count without issues?

**Comment by walkor** [comment]
Multiple processes cannot share the same database instance(or non_threadsafe_component), which may cause unpredictable problems.
The database instance should be instantiated separately in each process. For example, it should be initialized when onworkerstart or when the child process needs to operate the database.

**Comment by cayolblake** [comment]
@walkor so why is it working ? 🤔 

I verified from logging inside db first entry point that it's being initialized only one time when having count set to note than 1.

Is creating multiple workers similar to having multiple *individual* processes?

**Comment by walkor** [comment]
If multiple processes share a database connection, if one process closes the database connection, other processes will be affected. 
It is the right choice for each process to initialize its own database connection.

**Comment by cayolblake** [comment]
Non of the worker processes have the power to close connection.

Is that the only issue?

**Comment by walkor** [comment]
I don't think it's the only issue. There should be other issues I don't know

**Comment by joanhey** [comment]
If you use only 1 db connection for all your worker processes (and child processes), you have a bottleneck.

So performance will be a big issue.

**Comment by cayolblake** [comment]
@joanhey I did mention it's non-thread safe.
